File: app/src/main/java/net/ivpn/client/common/prefs/NetworkProtectionPreference.java
package net.ivpn.client.common.prefs;

/*
 IVPN Android app
 https://github.com/ivpn/android-app

 Created by Oleksandr Mykhailenko.
 Copyright (c) 2020 Privatus Limited.

 This file is part of the IVPN Android app.

 The IVPN Android app is free software: you can redistribute it and/or
 modify it under the terms of the GNU General Public License as published by the Free
 Software Foundation, either version 3 of the License, or (at your option) any later version.

 The IVPN Android app is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License
 along with the IVPN Android app. If not, see <https://www.gnu.org/licenses/>.
*/

import android.content.SharedPreferences;

import net.ivpn.client.vpn.model.NetworkState;

import java.util.HashSet;
import java.util.Set;

import javax.inject.Inject;

public class NetworkProtectionPreference {

    private static final String SETTINGS_TRUSTED_WIFI_LIST = "SETTINGS_TRUSTED_WIFI_LIST";
    private static final String SETTINGS_UNTRUSTED_WIFI_LIST = "SETTINGS_UNTRUSTED_WIFI_LIST";
    private static final String SETTINGS_NONE_WIFI_LIST = "SETTINGS_NONE_WIFI_LIST";

    private static final String SETTINGS_DEFAULT_NETWORK_STATE = "SETTINGS_DEFAULT_NETWORK_STATE";
    private static final String SETTINGS_MOBILE_DATE_NETWORK_STATE = "SETTINGS_MOBILE_DATE_NETWORK_STATE";

    private Preference preference;

    @Inject
    public NetworkProtectionPreference(Preference preference) {
        this.preference = preference;
    }

    public void markWifiAsTrusted(String wifiSsid) {
        Set<String> trustedWifi = getTrustedWifiList();
        if (wifiSsid == null || trustedWifi.contains(wifiSsid)) {
            return;
        }
        Set<String> newTrustedWifiList = new HashSet<>(trustedWifi);
        newTrustedWifiList.add(wifiSsid);

        SharedPreferences sharedPreferences = preference.getNetworkRulesSharedPreferences();
        sharedPreferences.edit()
                .putStringSet(SETTINGS_TRUSTED_WIFI_LIST, newTrustedWifiList)
                .apply();
    }

    public void removeMarkWifiAsTrusted(String wifiSsid) {
        Set<String> trustedWifi = getTrustedWifiList();
        if (wifiSsid == null || !trustedWifi.contains(wifiSsid)) {
            return;
        }
        Set<String> newTrustedWifiList = new HashSet<>(trustedWifi);
        newTrustedWifiList.remove(wifiSsid);

        SharedPreferences sharedPreferences = preference.getNetworkRulesSharedPreferences();
        sharedPreferences.edit()
                .putStringSet(SETTINGS_TRUSTED_WIFI_LIST, newTrustedWifiList)
                .apply();
    }

    public Set<String> getTrustedWifiList() {
        SharedPreferences sharedPreferences = preference.getNetworkRulesSharedPreferences();
        return sharedPreferences.getStringSet(SETTINGS_TRUSTED_WIFI_LIST, new HashSet<String>());
    }

    public void markWifiAsUntrusted(String wifiSsid) {
        Set<String> untrustedWifi = getUntrustedWifiList();
        if (wifiSsid == null || untrustedWifi.contains(wifiSsid)) {
            return;
        }
        Set<String> newTrustedWifiList = new HashSet<>(untrustedWifi);
        newTrustedWifiList.add(wifiSsid);

        SharedPreferences sharedPreferences = preference.getNetworkRulesSharedPreferences();
        sharedPreferences.edit()
                .putStringSet(SETTINGS_UNTRUSTED_WIFI_LIST, newTrustedWifiList)
                .apply();
    }

    public void removeMarkWifiAsUntrusted(String wifiSsid) {
        Set<String> untrustedWifi = getUntrustedWifiList();
        if (wifiSsid == null || !untrustedWifi.contains(wifiSsid)) {
            return;
        }
        Set<String> newTrustedWifiList = new HashSet<>(untrustedWifi);
        newTrustedWifiList.remove(wifiSsid);

        SharedPreferences sharedPreferences = preference.getNetworkRulesSharedPreferences();
        sharedPreferences.edit()
                .putStringSet(SETTINGS_UNTRUSTED_WIFI_LIST, newTrustedWifiList)
                .apply();
    }

    public Set<String> getUntrustedWifiList() {
        SharedPreferences sharedPreferences = preference.getNetworkRulesSharedPreferences();
        return sharedPreferences.getStringSet(SETTINGS_UNTRUSTED_WIFI_LIST, new HashSet<String>());
    }

    public void markWifiAsNone(String wifiSsid) {
        Set<String> noneWifis = getNoneWifiList();
        if (wifiSsid == null || noneWifis.contains(wifiSsid)) {
            return;
        }
        Set<String> newTrustedWifiList = new HashSet<>(noneWifis);
        newTrustedWifiList.add(wifiSsid);

        SharedPreferences sharedPreferences = preference.getNetworkRulesSharedPreferences();
        sharedPreferences.edit()
                .putStringSet(SETTINGS_NONE_WIFI_LIST, newTrustedWifiList)
                .apply();
    }

    public void removeMarkWifiAsNone(String wifiSsid) {
        Set<String> noneWifis = getNoneWifiList();
        if (wifiSsid == null || !noneWifis.contains(wifiSsid)) {
            return;
        }
        Set<String> newTrustedWifiList = new HashSet<>(noneWifis);
        newTrustedWifiList.remove(wifiSsid);

        SharedPreferences sharedPreferences = preference.getNetworkRulesSharedPreferences();
        sharedPreferences.edit()
                .putStringSet(SETTINGS_NONE_WIFI_LIST, newTrustedWifiList)
                .apply();
    }

    public Set<String> getNoneWifiList() {
        SharedPreferences sharedPreferences = preference.getNetworkRulesSharedPreferences();
        return sharedPreferences.getStringSet(SETTINGS_NONE_WIFI_LIST, new HashSet<>());
    }

    public void setDefaultNetworkState(NetworkState defaultNetworkState) {
        SharedPreferences sharedPreferences = preference.getNetworkRulesSharedPreferences();
        sharedPreferences.edit()
                .putString(SETTINGS_DEFAULT_NETWORK_STATE, defaultNetworkState.name())
                .apply();
    }

    public NetworkState getDefaultNetworkState() {
        SharedPreferences sharedPreferences = preference.getNetworkRulesSharedPreferences();
        String defaultNetworkStateName = sharedPreferences.getString(SETTINGS_DEFAULT_NETWORK_STATE,
                NetworkState.NONE.name());
        return NetworkState.valueOf(defaultNetworkStateName);
    }

    public void setMobileDataNetworkState(NetworkState mobileDataNetworkState) {
        SharedPreferences sharedPreferences = preference.getNetworkRulesSharedPreferences();
        sharedPreferences.edit()
                .putString(SETTINGS_MOBILE_DATE_NETWORK_STATE, mobileDataNetworkState.name())
                .apply();
    }

    public NetworkState getMobileDataNetworkState() {
        SharedPreferences sharedPreferences = preference.getNetworkRulesSharedPreferences();
        String mobileDataNetworkStateName = sharedPreferences.getString(SETTINGS_MOBILE_DATE_NETWORK_STATE,
                NetworkState.DEFAULT.name());
        return NetworkState.valueOf(mobileDataNetworkStateName);
    }

    public boolean isDefaultBehaviourExist() {
        SharedPreferences sharedPreferences = preference.getNetworkRulesSharedPreferences();
        return sharedPreferences.contains(SETTINGS_DEFAULT_NETWORK_STATE);
    }

    public boolean isMobileBehaviourExist() {
        SharedPreferences sharedPreferences = preference.getNetworkRulesSharedPreferences();
        return sharedPreferences.contains(SETTINGS_MOBILE_DATE_NETWORK_STATE);
    }
}

File: core/src/main/java/net/ivpn/core/common/prefs/NetworkProtectionPreference.kt
package net.ivpn.core.common.prefs

/*
IVPN Android app
https://github.com/ivpn/android-app

Created by Oleksandr Mykhailenko.
Copyright (c) 2023 IVPN Limited.

This file is part of the IVPN Android app.

The IVPN Android app is free software: you can redistribute it and/or
modify it under the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option) any later version.

The IVPN Android app is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
details.

You should have received a copy of the GNU General Public License
along with the IVPN Android app. If not, see <https://www.gnu.org/licenses/>.
*/

import net.ivpn.core.common.dagger.ApplicationScope
import net.ivpn.core.vpn.model.NetworkState
import java.util.*
import javax.inject.Inject

@ApplicationScope
class NetworkProtectionPreference @Inject constructor(
        private val preference: Preference
) {
    companion object {
        private const val SETTINGS_TRUSTED_WIFI_LIST = "SETTINGS_TRUSTED_WIFI_LIST"
        private const val SETTINGS_UNTRUSTED_WIFI_LIST = "SETTINGS_UNTRUSTED_WIFI_LIST"
        private const val SETTINGS_NONE_WIFI_LIST = "SETTINGS_NONE_WIFI_LIST"
        private const val SETTINGS_DEFAULT_NETWORK_STATE = "SETTINGS_DEFAULT_NETWORK_STATE"
        private const val SETTINGS_MOBILE_DATE_NETWORK_STATE = "SETTINGS_MOBILE_DATE_NETWORK_STATE"
    }

    val trustedWifiList: Set<String>?
        get() {
            val sharedPreferences = preference.networkRulesSharedPreferences
            return sharedPreferences.getStringSet(SETTINGS_TRUSTED_WIFI_LIST, HashSet())
        }

    val untrustedWifiList: Set<String>?
        get() {
            val sharedPreferences = preference.networkRulesSharedPreferences
            return sharedPreferences.getStringSet(SETTINGS_UNTRUSTED_WIFI_LIST, HashSet())
        }
    val noneWifiList: Set<String>?
        get() {
            val sharedPreferences = preference.networkRulesSharedPreferences
            return sharedPreferences.getStringSet(SETTINGS_NONE_WIFI_LIST, HashSet())
        }
    var defaultNetworkState: NetworkState
        get() {
            val sharedPreferences = preference.networkRulesSharedPreferences
            val defaultNetworkStateName = sharedPreferences.getString(SETTINGS_DEFAULT_NETWORK_STATE,
                    NetworkState.NONE.name)
            return NetworkState.valueOf(defaultNetworkStateName!!)
        }
        set(defaultNetworkState) {
            val sharedPreferences = preference.networkRulesSharedPreferences
            sharedPreferences.edit()
                    .putString(SETTINGS_DEFAULT_NETWORK_STATE, defaultNetworkState.name)
                    .apply()
        }
    var mobileDataNetworkState: NetworkState
        get() {
            val sharedPreferences = preference.networkRulesSharedPreferences
            val mobileDataNetworkStateName = sharedPreferences.getString(SETTINGS_MOBILE_DATE_NETWORK_STATE,
                    NetworkState.DEFAULT.name)
            return NetworkState.valueOf(mobileDataNetworkStateName!!)
        }
        set(mobileDataNetworkState) {
            val sharedPreferences = preference.networkRulesSharedPreferences
            sharedPreferences.edit()
                    .putString(SETTINGS_MOBILE_DATE_NETWORK_STATE, mobileDataNetworkState.name)
                    .apply()
        }
    val isDefaultBehaviourExist: Boolean
        get() {
            val sharedPreferences = preference.networkRulesSharedPreferences
            return sharedPreferences.contains(SETTINGS_DEFAULT_NETWORK_STATE)
        }
    val isMobileBehaviourExist: Boolean
        get() {
            val sharedPreferences = preference.networkRulesSharedPreferences
            return sharedPreferences.contains(SETTINGS_MOBILE_DATE_NETWORK_STATE)
        }

    fun markWifiAsTrusted(wifiSsid: String?) {
        val trustedWifi = trustedWifiList
        if (wifiSsid == null || trustedWifi == null || trustedWifi.contains(wifiSsid)) {
            return
        }
        val newTrustedWifiList: MutableSet<String> = HashSet(trustedWifi)
        newTrustedWifiList.add(wifiSsid)
        val sharedPreferences = preference.networkRulesSharedPreferences
        sharedPreferences.edit()
                .putStringSet(SETTINGS_TRUSTED_WIFI_LIST, newTrustedWifiList)
                .apply()
    }

    fun removeMarkWifiAsTrusted(wifiSsid: String?) {
        val trustedWifi = trustedWifiList
        if (wifiSsid == null || trustedWifi == null || !trustedWifi.contains(wifiSsid)) {
            return
        }

        val newTrustedWifiList: MutableSet<String> = HashSet(trustedWifi)
        newTrustedWifiList.remove(wifiSsid)
        val sharedPreferences = preference.networkRulesSharedPreferences
        sharedPreferences.edit()
                .putStringSet(SETTINGS_TRUSTED_WIFI_LIST, newTrustedWifiList)
                .apply()
    }

    fun markWifiAsUntrusted(wifiSsid: String?) {
        val untrustedWifi = untrustedWifiList
        if (wifiSsid == null || untrustedWifi == null || untrustedWifi.contains(wifiSsid)) {
            return
        }

        val newTrustedWifiList: MutableSet<String> = HashSet(untrustedWifi)
        newTrustedWifiList.add(wifiSsid)

        val sharedPreferences = preference.networkRulesSharedPreferences
        sharedPreferences.edit()
                .putStringSet(SETTINGS_UNTRUSTED_WIFI_LIST, newTrustedWifiList)
                .apply()
    }

    fun removeMarkWifiAsUntrusted(wifiSsid: String?) {
        val untrustedWifi = untrustedWifiList
        if (wifiSsid == null || untrustedWifi == null || !untrustedWifi.contains(wifiSsid)) {
            return
        }

        val newTrustedWifiList: MutableSet<String> = HashSet(untrustedWifi)
        newTrustedWifiList.remove(wifiSsid)

        val sharedPreferences = preference.networkRulesSharedPreferences
        sharedPreferences.edit()
                .putStringSet(SETTINGS_UNTRUSTED_WIFI_LIST, newTrustedWifiList)
                .apply()
    }

    fun markWifiAsNone(wifiSsid: String?) {
        val noneWiFis = noneWifiList
        if (wifiSsid == null || noneWiFis == null || noneWiFis.contains(wifiSsid)) {
            return
        }

        val newTrustedWifiList: MutableSet<String> = HashSet(noneWiFis)
        newTrustedWifiList.add(wifiSsid)

        val sharedPreferences = preference.networkRulesSharedPreferences
        sharedPreferences.edit()
                .putStringSet(SETTINGS_NONE_WIFI_LIST, newTrustedWifiList)
                .apply()
    }

    fun removeMarkWifiAsNone(wifiSsid: String?) {
        val noneWiFis = noneWifiList
        if (wifiSsid == null || noneWiFis == null || !noneWiFis.contains(wifiSsid)) {
            return
        }

        val newTrustedWifiList: MutableSet<String> = HashSet(noneWiFis)
        newTrustedWifiList.remove(wifiSsid)

        val sharedPreferences = preference.networkRulesSharedPreferences
        sharedPreferences.edit()
                .putStringSet(SETTINGS_NONE_WIFI_LIST, newTrustedWifiList)
                .apply()
    }
}